astrowidgets/glupyter.py

```diff
@@ -0,0 +1,400 @@
+        # multiple data into the same widget in the same session.
+        self._app.data_collection.clear()
+
+        self._app.add_datasets(data)
+        viewer.add_data(data)
+        viewer.state.reference_data = data  # Bring it to focus
+        self._link_image_to_cb()
+
+    def load_nddata(self, nddata, label=None):
+        """Load a `~astropy.nddata.NDData` object into the viewer.
+
+        .. todo:: Add flag/masking support, etc.
+
+        Parameters
+        ----------
+        nddata : `~astropy.nddata.NDData`
+            ``NDData`` with image data and WCS.
+
+        label : str or `None`
+            Glue label name for the data.
+
+        """
+        if label is None:
+            label = _gen_random_label()
+
+        data = Data(label=label)
+
+        if nddata.wcs is not None:
+            coords = coordinates_from_wcs(nddata.wcs)
+            data.coords = coords
+
+        data.add_component(component=nddata.data, label=label)
+
+        viewer = self._viewer
+        data = validate_data_argument(self._app.data_collection, data)
+
+        # NOTE: Changing state via API does not work properly if we loaded
+        # multiple data into the same widget in the same session.
+        self._app.data_collection.clear()
+
+        self._app.add_datasets(data)
+        viewer.add_data(data)
+        viewer.state.reference_data = data  # Bring it to focus
+        self._link_image_to_cb()
+
+    def load_array(self, arr, label=None):
+        """Load a 2D array into the viewer.
+
+        .. note:: Use :meth:`load_nddata` for WCS support.
+
+        Parameters
+        ----------
+        arr : array-like
+            2D array.
+
+        label : str or `None`
+            Glue label name for the data.
+
+        """
+        if label is None:
+            label = _gen_random_label()
+
+        data = Data(label=label)
+        data.add_component(component=arr, label=label)
+
+        viewer = self._viewer
+        data = validate_data_argument(self._app.data_collection, data)
+
+        # NOTE: Changing state via API does not work properly if we loaded
+        # multiple data into the same widget in the same session.
+        self._app.data_collection.clear()
+
+        self._app.add_datasets(data)
+        viewer.add_data(data)
+        viewer.state.reference_data = data  # Bring it to focus
+        self._link_image_to_cb()
+
+    def center_on(self, point):
+        raise NotImplementedError  # FIXME
+
+    def offset_to(self, dx, dy, skycoord_offset=False):
+        raise NotImplementedError  # FIXME
+
+    @property
+    def zoom_level(self):
+        raise NotImplementedError  # FIXME
+
+    @zoom_level.setter
+    def zoom_level(self, value):
+        raise NotImplementedError  # FIXME
+
+    # Need this here because we need to overwrite the setter.
+    @property
+    def marker(self):
+        return super().marker
+
+    @marker.setter
+    def marker(self, value):
+        raise NotImplementedError  # FIXME
```

Comment:
For all the marker stuff you could do it two ways - keep track of the markers in the astrowidgets layer and use the bqplot ``Scatter`` class to show the markers (you can add/remove markers by setting the ``x`` and ``y`` properties on the scatter object). I can show you an example if needed. Another option would be to actually make the list of markers be a glue data object and add it to the image viewer as a normal glue dataset - the advantage of this is that it shows up in the image viewer layers and you can control the appearance of the markers. There is a glue message you can fire if you want to update the glue dataset to force the viewer to refresh so let me know if you decide to go that route.

Reply:
I think the latter makes more sense. Currently in the Ginga implementation, it creates Ginga canvas objects, and then extracts them back from the Ginga canvas, so this is consistent.

What isn't clear to me is how you control the looks of the Glue markers. For example, how do I ask Glue to draw blue X or red circle? How do I only extract markers that are red circle? How do I extract all the markers irrespective of shape/color?